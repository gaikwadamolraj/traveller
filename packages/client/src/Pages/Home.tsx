import { useReducer, useRef, useState } from 'react'
import type { Reducer } from 'react'
import type { FC } from 'react'
import { Container, InputRightElement, Input, Heading, InputGroup, IconButton, VStack } from '@chakra-ui/react'
import { Search2Icon } from '@chakra-ui/icons'
import { TablePagination } from '../components/TablePagination'
import type { IAction, ICity, PutPayload, IState } from '../dataStructure'
import { CONSTANTS } from '../dataStructure'
import { sendApiCall } from '../httputil/Api'
import { Loading } from '../components/Loading'
import { Error } from '../components/Error'
import { initialState, reducer, setError } from '../store'

export const Home: FC = () => {
  const [pageNumber, setPage] = useState<number>(CONSTANTS.DEFAULT_PAGE)
  const [serachName, setSearchName] = useState<string>('')

  // eslint-disable-next-line
  const serachInput = useRef<any>()

  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)
  const { cities, isLoading, error } = state

  const fetchData = () => {
    const name: string = serachInput?.current?.value.trim()
    setSearchName(name)
    if (name) {
      sendApiCall({ dispatch, urlData: `?name=${name}` })
      setPage(CONSTANTS.DEFAULT_PAGE)
    } else {
      setError(dispatch, 'Please enter some char values')
    }
  }
  const updateData = async (city: ICity, page: number, payload: PutPayload) => {
    sendApiCall({ method: 'PUT', dispatch, urlData: `/${city.id}`, payload })
    setPage(page)
  }

  const handleVisited = (city: ICity, page: number) => {
    updateData(city, page, { visited: !city.visited })
  }

  const handleWishlist = (city: ICity, page: number) => {
    updateData(city, page, { wishlist: !city.wishlist })
  }

  // useEffect(() => {
  //   // fetchData()
  //   return () => {
  //     // Clean up the subscription if any
  //   }
  // }, [])

  return (
    <VStack>
      <Heading as="h1">Smart traveller</Heading>
      <Container maxW="container.md">
        <InputGroup>
          <Input
            placeholder={'Find city by name'}
            ref={serachInput}
            onKeyPress={e => (e.key === 'Enter' ? fetchData() : '')}
          />
          <InputRightElement children={<IconButton aria-label="" icon={<Search2Icon />} />} onClick={fetchData} />
        </InputGroup>
        <Error error={error} />

        {isLoading ? (
          <Loading />
        ) : serachName ? (
          <TablePagination
            pageNumber={pageNumber}
            cities={cities || []}
            handleVisited={handleVisited}
            handleWishlist={handleWishlist}
          />
        ) : (
          ''
        )}
      </Container>
    </VStack>
  )
}
