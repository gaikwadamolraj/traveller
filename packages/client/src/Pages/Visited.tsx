import { useEffect, useReducer } from 'react'
import type { Reducer } from 'react'
import { Container, Heading } from '@chakra-ui/react'
import type { IAction, IEmptyProps, IState } from '../dataStructure'
import { TablePagination } from '../components/TablePagination'
import { Loading } from '../components/Loading'
import { sendApiCall } from '../httputil/Api'
import { Error } from '../components/Error'
import { initialState, reducer } from '../store'

export const Visited: React.FC<IEmptyProps> = () => {
  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)
  const { cities, isLoading, error } = state

  useEffect(() => {
    sendApiCall({ dispatch, urlData: `/?visited=true` })
  }, [])

  return (
    <>
      <Heading as="h1">Visited</Heading>
      <Container maxW="container.md">
        <Error error={error} />
        {!isLoading ? <TablePagination cities={cities || []} isLoading /> : <Loading />}
      </Container>
    </>
  )
}
