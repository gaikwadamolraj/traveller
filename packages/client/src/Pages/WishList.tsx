import { useEffect, useReducer } from 'react'
import type { Reducer } from 'react'
import { Container, Heading } from '@chakra-ui/react'
import { TablePagination } from '../components/TablePagination'
import type { IAction, IEmptyProps, IState } from '../dataStructure'
import { sendApiCall } from '../httputil/Api'
import { Loading } from '../components/Loading'
import { Error } from '../components/Error'
import { initialState, reducer } from '../store'

export const WishList: React.FC<IEmptyProps> = () => {
  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)

  const { cities, isLoading, error } = state
  useEffect(() => {
    sendApiCall({ dispatch, urlData: `/?wishlist=true` })
  }, [])

  return (
    <>
      <Heading as="h1">Wish list</Heading>
      <Container maxW="container.md">
        <Error error={error} />

        {!isLoading ? <TablePagination cities={cities || []} isLoading /> : <Loading />}
      </Container>
    </>
  )
}
