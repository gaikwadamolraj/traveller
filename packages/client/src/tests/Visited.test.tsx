
import { Visited } from '../Pages/Visited'
import { render } from '../test-utils'

describe('Visited Page component', () => {
  it('Visited page', () => {
    const { container } = render(<Visited />)
    expect(container).toMatchSnapshot()
  })
})
