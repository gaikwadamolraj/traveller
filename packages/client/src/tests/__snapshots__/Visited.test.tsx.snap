// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visited Page component Visited page 1`] = `
<div>
  <h1
    class="chakra-heading css-1dklj6k"
  >
    Visited
  </h1>
  <div
    class="chakra-container css-nm5t63"
  >
    <div
      class="chakra-spinner css-17lgtxu"
    >
      <span
        class="css-f8n5zr"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
